src/transformers/models/bert/modeling_flax_bert.py

```diff
@@ -423,14 +363,22 @@ def __call__(self, hidden_states):
 class FlaxBertLMPredictionHead(nn.Module):
     config: BertConfig
     dtype: jnp.dtype = jnp.float32
+    bias_init: Callable[..., np.ndarray] = jax.nn.initializers.zeros
 
     def setup(self):
         self.transform = FlaxBertPredictionHeadTransform(self.config, dtype=self.dtype)
-        self.decoder = nn.Dense(self.config.vocab_size, dtype=self.dtype)
+        self.decoder = nn.Dense(self.config.vocab_size, dtype=self.dtype, use_bias=False)
+        self.bias = self.param("bias", self.bias_init, (self.config.vocab_size,))
 
-    def __call__(self, hidden_states):
+    def __call__(self, hidden_states, shared_embedding=None):
         hidden_states = self.transform(hidden_states)
-        hidden_states = self.decoder(hidden_states)
+
+        if shared_embedding is not None:
+            hidden_states = self.decoder.apply({"params": {"kernel": shared_embedding.T}}, hidden_states)
```

Comment:
Ok, I think this could be the way to go for weight sharing - it's different from the PyTorch design where the weights are shared before the forward pass & therefore there is less code duplication in PyTorch as we can have a general function for it. 

If I understood correctly this is more difficult in Flax, so I think this is a good solution for now. It will mean that we'll re-write this code in all models that have weight sharing, but on the bright side it's also arguably less "magic" and the user has an easier time understanding that weights are shared when tracing a forward pass.

=> What do you think @sgugger @CoderPat @avital @marcvanzee ?

Reply:
It looks like a workable solution, however this will lead to creation of an unused kernel for the `self.decoder` right? Also why not have the shared embedding as a parameter rather than argument to the forward pass (similar to [the example that I mentioned](https://github.com/google/flax/blob/a6282dba2f2be8d1724ea35f6e1d963097058cd9/examples/wmt/models.py#L342))? Is it because it would screw up the names and therefor the automatic pytorch checkpoint conversion?

Reply:
Yeah, I took this approach from the solution proposed [here](https://github.com/google/flax/discussions/1264#discussioncomment-644887) . The problem mostly is that we are quite restricted by the modularization structure to match PyTorch's ones. If I understand correctly, for your approach to work and for your approach to work I think we would have to create the layer outside of the word embedding module which doesn't work well with PyTorch's weight structure

Reply:
Also, (if I understand correctly), there are no kernel weights created because the code actually never traces through the model at init -> so this should be fine in terms of memory

Reply:
Ok then I agree this seems to best way to do this. 
In the future if Flax adds something that allows weight sharing between more two `nn.module` maybe it might be worth for you guys to think about changing it